File: server/app/models/game.py
# app/models/game.py
from typing import Optional, List
from enum import Enum
from app.models.board import Board
from app.models.position import Position
from app.models.piece import Color, Piece, PieceType, King, Rook, Pawn

class GameStatus(Enum):
    ACTIVE = "active"
    CHECKMATE = "checkmate"
    STALEMATE = "stalemate"
    DRAW = "draw"

class Move:
    """Represents a chess move"""
    def __init__(
        self, 
        from_pos: Position, 
        to_pos: Position,
        promotion_piece: Optional[PieceType] = None
    ):
        self.from_pos = from_pos
        self.to_pos = to_pos
        self.promotion_piece = promotion_piece
        self.captured_piece: Optional[Piece] = None
        self.is_castling = False
        self.is_en_passant = False
    
    def __str__(self) -> str:
        return f"{self.from_pos} to {self.to_pos}"
    
    def __repr__(self) -> str:
        return f"Move({self.from_pos}, {self.to_pos})"

class Game:
    """Main game class that handles all chess logic"""
    
    def __init__(self):
        self.board = Board()
        self.board.setup_initial_position()
        self.current_turn = Color.WHITE
        self.move_history: List[Move] = []
        self.status = GameStatus.ACTIVE
    
    @property
    def last_move(self) -> Optional[Move]:
        """Get the last move played"""
        return self.move_history[-1] if self.move_history else None
    
    def get_legal_moves(self, position: Position) -> List[Position]:
        """Get all legal moves for the piece at the given position"""
        piece = self.board.get_piece(position)
        if not piece or piece.color != self.current_turn:
            return []
        
        # Get possible moves from the piece
        possible_moves = piece.get_possible_moves(position, self.board)
        
        # Add special moves
        if isinstance(piece, King):
            possible_moves.extend(self._get_castling_moves(position))
        elif isinstance(piece, Pawn):
            possible_moves.extend(self._get_en_passant_moves(position))
        
        # Filter out moves that would leave king in check
        legal_moves = []
        for move_pos in possible_moves:
            if self._is_legal_move(position, move_pos):
                legal_moves.append(move_pos)
        
        return legal_moves
    
    def _is_legal_move(self, from_pos: Position, to_pos: Position) -> bool:
        """Check if a move is legal (doesn't leave king in check)"""
        # Make a temporary board to test the move
        test_board = self.board.clone()
        test_board.move_piece(from_pos, to_pos)
        
        # Handle en passant capture on test board
        piece = self.board.get_piece(from_pos)
        if isinstance(piece, Pawn):
            if abs(to_pos.col - from_pos.col) == 1 and self.board.get_piece(to_pos) is None:
                # En passant - remove the captured pawn
                test_board.set_piece(Position(from_pos.row, to_pos.col), None)
        
        return not test_board.is_in_check(self.current_turn)
    
    def make_move(
        self, 
        from_pos: Position, 
        to_pos: Position,
        promotion_piece: Optional[PieceType] = None
    ) -> bool:
        """
        Make a move. Returns True if successful, False otherwise.
        """
        if self.status != GameStatus.ACTIVE:
            return False
        
        piece = self.board.get_piece(from_pos)
        if not piece or piece.color != self.current_turn:
            return False
        
        legal_moves = self.get_legal_moves(from_pos)
        if to_pos not in legal_moves:
            return False
        
        # Create move object
        move = Move(from_pos, to_pos, promotion_piece)
        
        # Handle special moves
        if isinstance(piece, King) and abs(to_pos.col - from_pos.col) == 2:
            self._perform_castling(from_pos, to_pos)
            move.is_castling = True
        elif isinstance(piece, Pawn) and abs(to_pos.col - from_pos.col) == 1 and self.board.get_piece(to_pos) is None:
            self._perform_en_passant(from_pos, to_pos)
            move.is_en_passant = True
        else:
            # Regular move
            move.captured_piece = self.board.move_piece(from_pos, to_pos)
        
        # Handle pawn promotion
        if isinstance(piece, Pawn):
            promotion_row = 0 if piece.color == Color.WHITE else 7
            if to_pos.row == promotion_row:
                self._promote_pawn(to_pos, promotion_piece or PieceType.QUEEN)
        
        # Add to history
        self.move_history.append(move)
        
        # Switch turns
        self.current_turn = self.current_turn.opposite()
        
        # Update game status
        self._update_game_status()
        
        return True
    
    def _get_castling_moves(self, king_pos: Position) -> List[Position]:
        """Get castling moves for the king"""
        moves = []
        king = self.board.get_piece(king_pos)
        
        if not isinstance(king, King) or king.has_moved:
            return moves
        
        if self.board.is_in_check(self.current_turn):
            return moves
        
        back_rank = 7 if self.current_turn == Color.WHITE else 0
        
        # Kingside castling
        kingside_rook = self.board.get_piece(Position(back_rank, 7))
        if isinstance(kingside_rook, Rook) and not kingside_rook.has_moved:
            if (self.board.get_piece(Position(back_rank, 5)) is None and
                self.board.get_piece(Position(back_rank, 6)) is None):
                if (not self.board.is_square_under_attack(Position(back_rank, 5), self.current_turn.opposite()) and
                    not self.board.is_square_under_attack(Position(back_rank, 6), self.current_turn.opposite())):
                    moves.append(Position(back_rank, 6))
        
        # Queenside castling
        queenside_rook = self.board.get_piece(Position(back_rank, 0))
        if isinstance(queenside_rook, Rook) and not queenside_rook.has_moved:
            if (self.board.get_piece(Position(back_rank, 1)) is None and
                self.board.get_piece(Position(back_rank, 2)) is None and
                self.board.get_piece(Position(back_rank, 3)) is None):
                if (not self.board.is_square_under_attack(Position(back_rank, 2), self.current_turn.opposite()) and
                    not self.board.is_square_under_attack(Position(back_rank, 3), self.current_turn.opposite())):
                    moves.append(Position(back_rank, 2))
        
        return moves
    
    def _perform_castling(self, king_pos: Position, king_target: Position) -> None:
        """Perform castling move"""
        back_rank = king_pos.row
        
        # Move king
        self.board.move_piece(king_pos, king_target)
        
        # Move rook
        if king_target.col == 6:  # Kingside
            rook_from = Position(back_rank, 7)
            rook_to = Position(back_rank, 5)
            self.board.move_piece(rook_from, rook_to)
        else:  # Queenside
            rook_from = Position(back_rank, 0)
            rook_to = Position(back_rank, 3)
            self.board.move_piece(rook_from, rook_to)
    
    def _get_en_passant_moves(self, pawn_pos: Position) -> List[Position]:
        """Get en passant capture moves for a pawn"""
        moves = []
        
        if not self.last_move:
            return moves
        
        pawn = self.board.get_piece(pawn_pos)
        if not isinstance(pawn, Pawn):
            return moves
        
        en_passant_row = 3 if pawn.color == Color.WHITE else 4
        if pawn_pos.row != en_passant_row:
            return moves
        
        # Check if last move was a two-square pawn move next to our pawn
        last_piece = self.board.get_piece(self.last_move.to_pos)
        if (isinstance(last_piece, Pawn) and
            abs(self.last_move.from_pos.row - self.last_move.to_pos.row) == 2 and
            self.last_move.to_pos.row == pawn_pos.row and
            abs(self.last_move.to_pos.col - pawn_pos.col) == 1):
            
            direction = -1 if pawn.color == Color.WHITE else 1
            capture_pos = Position(pawn_pos.row + direction, self.last_move.to_pos.col)
            moves.append(capture_pos)
        
        return moves
    
    def _perform_en_passant(self, from_pos: Position, to_pos: Position) -> None:
        """Perform en passant capture"""
        self.board.move_piece(from_pos, to_pos)
        # Remove the captured pawn
        self.board.set_piece(Position(from_pos.row, to_pos.col), None)
    
    def _promote_pawn(self, position: Position, piece_type: PieceType) -> None:
        """Promote a pawn to another piece"""
        pawn = self.board.get_piece(position)
        if not isinstance(pawn, Pawn):
            return
        
        # Import here to avoid circular imports
        from app.models.piece import Queen, Rook, Bishop, Knight
        
        piece_classes = {
            PieceType.QUEEN: Queen,
            PieceType.ROOK: Rook,
            PieceType.BISHOP: Bishop,
            PieceType.KNIGHT: Knight,
        }
        
        piece_class = piece_classes.get(piece_type, Queen)
        new_piece = piece_class(pawn.color, has_moved=True)
        self.board.set_piece(position, new_piece)
    
    def _update_game_status(self) -> None:
        """Update game status (check for checkmate/stalemate)"""
        # Check if current player has any legal moves
        has_legal_moves = False
        for pos, piece in self.board.get_all_pieces(self.current_turn):
            if self.get_legal_moves(pos):
                has_legal_moves = True
                break
        
        if not has_legal_moves:
            if self.board.is_in_check(self.current_turn):
                self.status = GameStatus.CHECKMATE
            else:
                self.status = GameStatus.STALEMATE
    
    def is_checkmate(self) -> bool:
        """Check if current position is checkmate"""
        return self.status == GameStatus.CHECKMATE
    
    def is_stalemate(self) -> bool:
        """Check if current position is stalemate"""
        return self.status == GameStatus.STALEMATE
    
    def is_check(self) -> bool:
        """Check if current player is in check"""
        return self.board.is_in_check(self.current_turn)
    
    def display(self) -> str:
        """Display the current game state"""
        lines = [self.board.display()]
        lines.append(f"\nCurrent turn: {self.current_turn.value}")
        lines.append(f"Status: {self.status.value}")
        
        if self.is_check():
            lines.append("CHECK!")
        
        return "\n".join(lines)

File: client/src/game/game.ts
import { Board } from "./board";
import { Position } from "./position";
import { Move } from "./move";
import { Piece, Pawn, King, Queen, Rook, Bishop, Knight } from "./piece";
import type {
  Color,
  PieceType,
  GameStatus,
  PositionData,
  HistoryData,
} from "../types/chess_types";
import type { GameHistory } from "./types";

export class Game {
  initialFEN: string;
  board: Board;
  currentTurn: Color;
  moveHistory: Move[];
  status: GameStatus;
  gameHistory: GameHistory[];
  currentHistoryIndex: number;

  constructor(fen: string = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR") {
    this.initialFEN = fen;
    this.board = new Board(fen);
    this.currentTurn = "white";
    this.moveHistory = [];
    this.status = "active";
    this.gameHistory = [
      {
        status: this.status,
        board: this.board.clone(),
      },
    ];
    this.currentHistoryIndex = 0;
  }

  get lastMove(): Move | undefined {
    if (this.currentHistoryIndex === 0) return undefined;
    return this.moveHistory[this.currentHistoryIndex - 1];
  }

  getLegalMoves(position: Position): Position[] {
    const piece = this.board.getPiece(position);
    if (!piece || piece.color !== this.currentTurn) return [];

    const possibleMoves = piece.getPossibleMoves(position, this.board);

    if (piece instanceof King)
      possibleMoves.push(...this.getCastlingMoves(position));
    else if (piece instanceof Pawn)
      possibleMoves.push(...this.getEnPassantMoves(position));

    return possibleMoves.filter((movePos) =>
      this.isLegalMove(position, movePos)
    );
  }

  serializeLegalMoves(position: Position): PositionData[] {
    const legalMoves = this.getLegalMoves(position);
    return legalMoves.map((pos) => pos.serializePosition());
  }

  private isLegalMove(fromPos: Position, toPos: Position): boolean {
    const testBoard = this.board.clone();
    testBoard.movePiece(fromPos, toPos);

    const piece = this.board.getPiece(fromPos);
    if (piece instanceof Pawn) {
      if (
        Math.abs(toPos.col - fromPos.col) === 1 &&
        !this.board.getPiece(toPos)
      ) {
        // En passant capture
        testBoard.setPiece(new Position(fromPos.row, toPos.col), null);
      }
    }

    return !testBoard.isInCheck(this.currentTurn);
  }

  makeMove(
    fromPos: Position,
    toPos: Position,
    promotionPiece?: PieceType
  ): boolean {
    const piece = this.board.getPiece(fromPos);
    if (!piece || piece.color !== this.currentTurn) return false;

    const legalMoves = this.getLegalMoves(fromPos);
    if (!legalMoves.some((p) => p.row === toPos.row && p.col === toPos.col))
      return false;

    // if not at the end of history truncate future moves
    if (this.currentHistoryIndex < this.gameHistory.length - 1) {
      this.gameHistory = this.gameHistory.slice(
        0,
        this.currentHistoryIndex + 1
      );
      this.moveHistory = this.moveHistory.slice(0, this.currentHistoryIndex);
    }

    const move = new Move(fromPos, toPos, promotionPiece);
    move.movingPiece = Object.assign(
      Object.create(Object.getPrototypeOf(piece)),
      piece
    );
    move.disambiguationInfo = this.getDisambiguationInfo(fromPos, toPos, piece);

    if (piece instanceof King && Math.abs(toPos.col - fromPos.col) === 2) {
      this.performCastling(fromPos, toPos);
      move.isCastling = true;
    }
    // en passant
    else if (
      piece instanceof Pawn &&
      Math.abs(toPos.col - fromPos.col) === 1 &&
      !this.board.getPiece(toPos)
    ) {
      this.performEnPassant(fromPos, toPos);
      move.isEnPassant = true;
    }
    // regular move
    else {
      move.capturedPiece = this.board.movePiece(fromPos, toPos);
    }

    if (piece instanceof Pawn) {
      const promotionRow = piece.color === "white" ? 0 : 7;
      if (toPos.row === promotionRow) {
        this.promotePawn(toPos, promotionPiece || "queen");
      }
    }

    this.currentTurn = this.currentTurn === "white" ? "black" : "white";
    this.moveHistory.push(move);
    this.gameHistory.push({ status: this.status, board: this.board.clone() });
    this.currentHistoryIndex = this.gameHistory.length - 1;

    this.updateGameStatus();

    move.isCheck = this.board.isInCheck(this.currentTurn);
    move.isCheckmate = this.status === "checkmate";

    return true;
  }

  jumpToMove(index: number): boolean {
    if (index < 0 || index >= this.gameHistory.length) return false;

    this.currentHistoryIndex = index;
    const historyState = this.gameHistory[index];

    this.board = historyState.board.clone();
    this.status = historyState.status;

    this.currentTurn = index % 2 === 0 ? "white" : "black";

    return true;
  }

  undo(): boolean {
    if (this.currentHistoryIndex <= 0) return false;
    return this.jumpToMove(this.currentHistoryIndex - 1);
  }

  redo(): boolean {
    if (this.currentHistoryIndex >= this.gameHistory.length - 1) return false;
    return this.jumpToMove(this.currentHistoryIndex + 1);
  }

  canUndo(): boolean {
    return this.currentHistoryIndex > 0;
  }

  canRedo(): boolean {
    return this.currentHistoryIndex < this.gameHistory.length - 1;
  }

  isAtCurrentPosition(): boolean {
    return this.currentHistoryIndex === this.gameHistory.length - 1;
  }

  getCurrentMoveNumber(): number {
    return this.currentHistoryIndex;
  }

  getTotalMoves(): number {
    return this.gameHistory.length - 1;
  }

  getCurrentMove(): Move | null {
    if (this.currentHistoryIndex === 0) return null;
    return this.moveHistory[this.currentHistoryIndex - 1];
  }

  private getCastlingMoves(kingPos: Position): Position[] {
    const moves: Position[] = [];
    const king = this.board.getPiece(kingPos);
    if (!(king instanceof King) || king.hasMoved) return moves;
    if (this.board.isInCheck(this.currentTurn)) return moves;

    const backRank = this.currentTurn === "white" ? 7 : 0;

    // kingside
    const kingsideRook = this.board.getPiece(new Position(backRank, 7));
    if (kingsideRook instanceof Rook && !kingsideRook.hasMoved) {
      if (
        !this.board.getPiece(new Position(backRank, 5)) &&
        !this.board.getPiece(new Position(backRank, 6)) &&
        !this.board.isSquareUnderAttack(
          new Position(backRank, 5),
          this.currentTurn === "white" ? "black" : "white"
        ) &&
        !this.board.isSquareUnderAttack(
          new Position(backRank, 6),
          this.currentTurn === "white" ? "black" : "white"
        )
      ) {
        moves.push(new Position(backRank, 6));
      }
    }

    // queenside
    const queensideRook = this.board.getPiece(new Position(backRank, 0));
    if (queensideRook instanceof Rook && !queensideRook.hasMoved) {
      if (
        !this.board.getPiece(new Position(backRank, 1)) &&
        !this.board.getPiece(new Position(backRank, 2)) &&
        !this.board.getPiece(new Position(backRank, 3)) &&
        !this.board.isSquareUnderAttack(
          new Position(backRank, 2),
          this.currentTurn === "white" ? "black" : "white"
        ) &&
        !this.board.isSquareUnderAttack(
          new Position(backRank, 3),
          this.currentTurn === "white" ? "black" : "white"
        )
      ) {
        moves.push(new Position(backRank, 2));
      }
    }

    return moves;
  }

  private performCastling(kingPos: Position, kingTarget: Position) {
    const backRank = kingPos.row;
    this.board.movePiece(kingPos, kingTarget);

    if (kingTarget.col === 6) {
      // kingside
      this.board.movePiece(
        new Position(backRank, 7),
        new Position(backRank, 5)
      );
    } else {
      // queenside
      this.board.movePiece(
        new Position(backRank, 0),
        new Position(backRank, 3)
      );
    }
  }

  private getEnPassantMoves(pawnPos: Position): Position[] {
    const moves: Position[] = [];
    const lastMove = this.lastMove;
    const pawn = this.board.getPiece(pawnPos);
    if (!lastMove || !(pawn instanceof Pawn)) return moves;

    const enPassantRow = pawn.color === "white" ? 3 : 4;
    if (pawnPos.row !== enPassantRow) return moves;

    const lastMovedPiece = this.board.getPiece(lastMove.toPos);
    if (
      lastMovedPiece instanceof Pawn &&
      Math.abs(lastMove.fromPos.row - lastMove.toPos.row) === 2 &&
      lastMove.toPos.row === pawnPos.row &&
      Math.abs(lastMove.toPos.col - pawnPos.col) === 1
    ) {
      const direction = pawn.color === "white" ? -1 : 1;
      moves.push(new Position(pawnPos.row + direction, lastMove.toPos.col));
    }

    return moves;
  }

  private performEnPassant(fromPos: Position, toPos: Position) {
    this.board.movePiece(fromPos, toPos);
    this.board.setPiece(new Position(fromPos.row, toPos.col), null);
  }

  private promotePawn(pos: Position, pieceType: PieceType) {
    const pawn = this.board.getPiece(pos);
    if (!(pawn instanceof Pawn)) return;

    const pieceMap: Record<
      PieceType,
      new (color: Color, hasMoved?: boolean) => Piece
    > = {
      queen: Queen,
      rook: Rook,
      bishop: Bishop,
      knight: Knight,
      king: King,
      pawn: Pawn,
    };

    const PieceClass = pieceMap[pieceType] || Queen;
    const newPiece = new PieceClass(pawn.color, true);
    this.board.setPiece(pos, newPiece);
  }

  private getDisambiguationInfo(
    fromPos: Position,
    toPos: Position,
    piece: Piece
  ): { needsFile?: boolean; needsRank?: boolean } {
    if (piece instanceof Pawn || piece instanceof King) {
      return {};
    }

    // find all pieces of the same type and color that can move to the same square
    const sameTypePieces = this.board
      .getAllPieces(piece.color)
      .filter(([pos, p]) => {
        if (p.pieceType !== piece.pieceType) return false;
        if (pos.row === fromPos.row && pos.col === fromPos.col) return false;

        const legalMoves = this.getLegalMoves(pos);
        return legalMoves.some(
          (m) => m.row === toPos.row && m.col === toPos.col
        );
      });

    if (sameTypePieces.length === 0) {
      return {};
    }

    // check if file disambiguation is enough
    const sameFile = sameTypePieces.some(([pos]) => pos.col === fromPos.col);
    const sameRank = sameTypePieces.some(([pos]) => pos.row === fromPos.row);

    if (!sameFile) {
      return { needsFile: true };
    } else if (!sameRank) {
      return { needsRank: true };
    } else {
      return { needsFile: true, needsRank: true };
    }
  }

  updateGameStatus() {
    let hasLegalMoves = false;
    for (const [pos] of this.board.getAllPieces(this.currentTurn)) {
      if (this.getLegalMoves(pos).length > 0) {
        hasLegalMoves = true;
        break;
      }
    }

    if (!hasLegalMoves) {
      this.status = this.board.isInCheck(this.currentTurn)
        ? "checkmate"
        : "stalemate";
    } else {
      this.status = "active";
    }
  }

  resetGame() {
    this.board = new Board(this.initialFEN);
    this.currentTurn = "white";
    this.moveHistory = [];
    this.status = "active";
    this.gameHistory = [
      {
        status: this.status,
        board: this.board.clone(),
      },
    ];
    this.currentHistoryIndex = 0;
  }

  isCheckmate(): boolean {
    return this.status === "checkmate";
  }

  isStalemate(): boolean {
    return this.status === "stalemate";
  }

  isCheck(): boolean {
    return this.board.isInCheck(this.currentTurn);
  }

  getMoveList(): string[] {
    return this.moveHistory.map((move) => move.toSAN());
  }

  serializeHistory(): HistoryData[] {
    return this.gameHistory.map((h) => ({
      status: h.status,
      board: h.board.serializeBoard(),
    }));
  }

  display(): string {
    const lines = [this.board.display()];
    lines.push(`\nCurrent turn: ${this.currentTurn}`);
    lines.push(`Status: ${this.status}`);
    if (this.isCheck()) lines.push("CHECK!");
    lines.push(`Move: ${this.currentHistoryIndex}/${this.getTotalMoves()}`);
    return lines.join("\n");
  }
}
